File: app/src/main/java/org/citra/citra_emu/utils/DiskShaderCacheProgress.java
package org.citra.citra_emu.utils;

import android.util.Log;

import org.citra.citra_emu.CitraHost;

/**
 * 着色器磁盘缓存加载进度回调（native → Java）。
 *
 * 上游在此显示进度 UI；NesStation 进程内模式只记录 + 轻量状态钩子。
 * static loadProgress(LoadCallbackStage,int,int,String) 被 native 直接调用，
 * 签名/静态性必须与上游一致。
 */
public final class DiskShaderCacheProgress {

    /** 阶段枚举（ordinal 被 native 使用 —— 顺序不可改）。 */
    public enum LoadCallbackStage {
        Prepare,
        Decompile,
        Build,
        Complete
    }

    public static final DiskShaderCacheProgress INSTANCE = new DiskShaderCacheProgress();

    private DiskShaderCacheProgress() {
    }

    public static void loadProgress(LoadCallbackStage stage, int current, int total, String obj) {
        Log.i("CitraHost", "shader cache " + stage + " " + current + "/" + total
            + (obj == null || obj.isEmpty() ? "" : " " + obj));
        CitraHost.StatusHook hook = CitraHost.getStatusHook();
        if (hook != null && stage == LoadCallbackStage.Complete) {
            hook.onStatus("着色器缓存就绪");
        }
    }
}

File: app/src/main/java/org/citra/citra_emu/utils/DiskShaderCacheProgress.kt
package org.citra.citra_emu.utils

import android.util.Log
import androidx.annotation.Keep

/**
 * Azahar 磁盘着色器缓存进度 —— JNI 契约桩。
 *
 * 原生侧在着色器编译期间通过 id_cache.cpp 调用
 * `loadProgress(Lorg/citra/citra_emu/utils/DiskShaderCacheProgress$LoadCallbackStage;IILjava/lang/String;)V`，
 * 并读取内嵌枚举 LoadCallbackStage 的静态字段 Prepare / Decompile / Build / Complete。
 * NesStation 用日志 + 简单状态输出代替原版进度对话框。
 */
@Keep
object DiskShaderCacheProgress {

    enum class LoadCallbackStage {
        Prepare,
        Decompile,
        Build,
        Complete
    }

    @JvmStatic
    fun loadProgress(stage: LoadCallbackStage, progress: Int, max: Int, obj: String) {
        Log.i("AzaharNative", "ShaderCache[$stage] $progress/$max $obj")
    }
}
